### PR TITLE
Support gpqa benchmark evaluation

### PR DESCRIPTION
**Pull request by 1jsingh**
**What does this PR do?**
This PR implements the evaluation harness on the recently released GPQA benchmark introduced in [GPQA: A Graduate-Level Google-Proof Q&A Benchmark](https://arxiv.org/abs/2308.07124). In particular, the code implements the evaluation of agents on the GPQA Benchmark with Open Book setting (discussed with @xingyaoww), where the agent can use external tools (e.g., python for calculation) and web-search to find relevant facts in order to answer the questions.

**Details of the benchmark**
- The benchmark consists of 448 high-quality and extremely difficult multiple-choice questions in the domains of biology, physics, and chemistry. The questions are intentionally designed to be "Google-proof," meaning that even highly skilled non-expert validators achieve only 34% accuracy despite unrestricted access to the web.
- Even experts in the corresponding domains achieve only 65% accuracy.
- State-of-the-art LLM models (e.g., GPT4) achieve only 39% accuracy on this challenging dataset.

**Why is it relevant in evaluating OpenDevin?**
Accurate solving of above graduate level questions would require both tool use (e.g., python for calculations) and web-search for finding related facts as information required for the questions might not be part of the LLM knowledge / training data.

Further references:
- https://arxiv.org/pdf/2311.12022
- https://paperswithcode.com/dataset/gpqa
- https://github.com/idavidrein/gpqa

**TODOs**
- [ ] Add support for other agents (currently only tested on `CodeActAgent`)
- [ ] Complete full benchmark evaluation with both gpt4, gpt4-turbo and gpt-4o
- [ ] Fix intermittent `BrowserException: Failed to start browser environment` error

### CODE REVIEW

**Comment by li-boxuan** [comment]
> Fix intermittent BrowserException: Failed to start browser environment error

May I know what opendevin version (commit hash) were you using? This might have been fixed very recently.

**Comment by 1jsingh** [comment]
9207a8da01dc6d874216edfac00bce31a6b40461

**Comment by li-boxuan** [comment]
Yeah plz try the latest commit. This was fixed in 9207a8 two days ago.

File: evaluation/gpqa/run_infer.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+        'metadata': metadata,
+        'history': [
+            (event_to_dict(action), event_to_dict(obs)) for action, obs in state.history
+        ],
+        'error': state.error if state and state.error else None,
+        'test_result': test_result,
+    }
+
+    config.workspace_mount_path = old_workspace_mount_path
+    config.workspace_base = old_workspace_base
```
Comment (li-boxuan) [inline evaluation/gpqa/run_infer.py:286]: Please apply a fix here, see https://github.com/OpenDevin/OpenDevin/pull/2214

**Review by li-boxuan — COMMENTED** [review]
Could you also please write a simple script that could extract the evaluation results? E.g. https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/swe_bench/scripts/summarise_results.py

File: evaluation/gpqa/run_infer.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+    output = {
+        'task_id': instance.task_id,
+        'instance_id': instance.instance_id,
+        'instruction': instruction,
+        'metadata': metadata,
+        'history': [
+            (event_to_dict(action), event_to_dict(obs)) for action, obs in state.history
+        ],
+        'error': state.error if state and state.error else None,
+        'test_result': test_result,
```
Comment (li-boxuan) [inline evaluation/gpqa/run_infer.py:282]: Please also add `metrics`, see https://github.com/OpenDevin/OpenDevin/pull/2199 for example

Comment (yufansong) [reply evaluation/gpqa/run_infer.py:282]: Yeah, pls add `metrics`, then we can get the average cost of each task which need in our final paper.

**Review by neubig — COMMENTED** [review]
Thanks @1jsingh , @li-boxuan , and @yufansong ! Now that we're mostly done with the paper it'd be great to get this merged in.

**Review by yufansong — APPROVED** [review]

**Comment by yufansong** [comment]
I do some fix. It should be good to merge now.
